### DISCUSSION TITLE
BaseServer.on() returning None instead of original handler

### DISCUSSION THREAD

**Discussion by dokterbob** [miguelgrinberg/python-socketio#1410] [General]
While working on [types](https://github.com/python/typeshed/pull/13095) for `BaseServer`'s `on()`, I noticed that the [original implementation](https://github.com/miguelgrinberg/python-socketio/blob/8f0e66c1cd1cd63dcef703576cc9cb9c99104df7/src/socketio/base_server.py#L125) returns `None` when used as a decorator. 

AFAIK, this is not the most common pattern: people might want to use the already-decorated value for decorating something else, only to find they can't. It also makes the types unnecessarily complex, over simply always returning the event handler callable.

Just to understand, I guess the `set_handler()` pattern allows for the following AFAIK undocumented behaviour:
```py
message_handler = sio.on('message')

@message_handler
def my_handler(): ...
```
After this, `my_handler()` will have it's original value, just the hook was added to `sio`'s handler.

In any other case, e.g. when `handler` is specified, it will _call_ `set_handler()` but not return it's return value. So:
```py
sio.on('message')
def my_handler(): ...
my_handler == None
```

Why has this decision been made, should the last line read `return set_handler(handler)`?

https://github.com/miguelgrinberg/python-socketio/blob/8f0e66c1cd1cd63dcef703576cc9cb9c99104df7/src/socketio/base_server.py#L127

Related: https://github.com/python/typeshed/pull/13095

**Comment by miguelgrinberg** [comment]
I think you are misunderstanding how the `on()` method works.

When used as a decorator, the `handler` argument is `None`. In that case `set_handler` is returned as the decorator's inner function. This function will then be invoked with the user's function to decorate as an argument.

When used as a regular method, you include the handler function in the call, so `handler` is not going to be `None`. In that case there is no need to return anything, since you are passing all the information directly. Example:

```python
sio.on('message', my_message_handler, namespace='/chat')
```

I reviewed the code again and it seems correct to me. Am I missing anything?
